### DISCUSSION TITLE
Reaction Matching with Reaction SMARTS help

### DISCUSSION THREAD

**Discussion by ebrowndev** [rdkit/rdkit#6954] [Q&A]
Hi Everyone,
I'm working on a script that uses reaction SMARTS to classify reactions (input as SMILES reaction strings), and I'm having some issues with certain SMARTS patterns I've written having some unexpected behavior. I believe it may come down to my misunderstanding what's going on behind the scenes when using rdkit.Chem.rdChemReactions.HasReactionSubstructMatch. 

Here's my definition for the Gatterman formylation, for example:
`[cX3H1:1].[CX2H:2]#[NX1:3] >> [c:1]-[C:2](=[O])`

And for the Friedel-Crafts Acylation:
`[cX3H1:1].[CX3:2](=[O:3])-[F,Cl,Br,I:4] >> [c:1]-[C:2](=[O:3])`

My understanding was that HasReactionSubstructMatch() will match multiple reactants if given in the pattern like those above. So in the Gatterman, I've defined the second reactant, hydrogen cyanide, and for the FC, the second reactant an acyl halide. So when querying for a reaction substructure match, it should only match one of those template strings when one of those specified reactant types are present, even if the product templates are the same. Unfortunately, it doesn't seem to be taking into account the structure of the second reactant at all, just it's presence. So when I do:

```
test_rxn = AllChem.ReactionFromSmarts("c1ccccc1.CCC(=O)Cl>>CCC(=O)c1ccccc1", {}, True)
gatterman_rxn = AllChem.ReactionFromSmarts("[cX3H1:1].[CX2H:2]#[NX1:3] >> [c:1]-[C:2](=[O])")
fc_rxn = AllChem.ReactionFromSmarts("[cX3H1:1].[CX3:2](=[O:3])-[F,Cl,Br,I:4] >> [c:1]-[C:2](=[O:3])")
Reactions.SanitizeRxn(gatterman_rxn)
Reactions.SanitizeRxn(fc_rxn)
Reactions.SanitizeRxn(test_rxn)

Reactions.HasReactionSubstructMatch(test_rxn, fc_rxn)
True
Reactions.HasReactionSubstructMatch(test_rxn, gatterman_rxn)
 True

```

But if I change the queried reaction, and remove it's second reactant:
```
test_rxn = AllChem.ReactionFromSmarts("c1ccccc1>>CCC(=O)c1ccccc1", {}, True)
Reactions.SanitizeRxn(test_rxn)

Reactions.HasReactionSubstructMatch(test_rxn, fc_rxn)
False
Reactions.HasReactionSubstructMatch(test_rxn, gatterman_rxn)
False
```

Interestingly, when I add seemingly any second reactant, they both return true again, so the presence of a second reactant is required, but no substructure matching seems to be done on the second reactant.
```
test_rxn = AllChem.ReactionFromSmarts("c1ccccc1.N>>CCC(=O)c1ccccc1", {}, True)
Reactions.SanitizeRxn(test_rxn)

Reactions.HasReactionSubstructMatch(test_rxn, fc_rxn)
True
Reactions.HasReactionSubstructMatch(test_rxn, gatterman_rxn)
True
```

So in short, why does my reaction SMARTS for the Gatterman formylation also match the FC-acylation, when there is no HCN or any other cyano group in the reactants in the reaction query? How do I write the queries such that both reactants are substructure matched? Is there a better approach for what I'm trying to accomplish?

**Comment by bp-kelley** [comment]
If looks like (after reviewing the code) the definition is the number of reactants has to be the same, and at lead one of the reactant templates has to match, the the order doesn't matter.  

```
>>> test_rxn = AllChem.ReactionFromSmarts("c1ccccc1>>CCC(=O)c1ccccc1", {}, True)
>>> AllChem.HasReactionSubstructMatch(test_rxn, fc_rxn)
False
```

```
>>> test_rxn = AllChem.ReactionFromSmarts("P.c1ccccc1>>CCC(=O)c1ccccc1", {}, True)
>>> AllChem.HasReactionSubstructMatch(test_rxn, fc_rxn)
True
```

It's easy to roll your own, btw:

```
test_rxn = AllChem.ReactionFromSmarts("c1ccccc1.C(=O)F>>CCC(=O)c1ccccc1", {}, True)
AllChem.SanitizeRxn(test_rxn)
for i, (t,q) in enumerate(zip(test_rxn.GetReactants(), fc_rxn.GetReactants())):
    print(f"Reactant {i} Matches {t.HasSubstructMatch(q, fc_rxn.GetSubstructParams())}")
```
